In [59]:
import pandas as pd
import numpy as np

brain_regis = pd.read_excel("/home/penglab/NMTcode/brain_registration_10092019_new.xlsx",index_col= [0], skiprows=[0], 
              names = ['ID','y_initial','x_initial','z_initial','y_after','x_after','z_after',
                       'ration','flip','flip_axis'])
brain_regis.loc[:,'x_downsample'] = brain_regis.loc[:,'x_initial']/brain_regis.loc[:,'x_after']
brain_regis.loc[:,'y_downsample'] = brain_regis.loc[:,'y_initial']/brain_regis.loc[:,'y_after']
brain_regis.loc[:,'z_downsample'] = brain_regis.loc[:,'z_initial']/brain_regis.loc[:,'z_after']
brain_regis.drop_duplicates(keep='first', inplace=True)
brain_regis.fillna(value='NA', inplace=True)
brain_regis =  brain_regis.loc[~brain_regis.index.duplicated(keep='first')]
brain_regis .loc[:, 'x_initial'] = brain_regis .loc[:, 'x_initial'].astype(float)
brain_regis .loc[:, 'x_initial'] = brain_regis .loc[:, 'x_initial'].astype(float)
brain_regis .loc[:, 'x_initial'] = brain_regis .loc[:, 'x_initial'].astype(float)

## Requirement
* Assume inside the main folder, each brain has its separrate folder, which includes
    * Folder **ori**
        * Storing resampled .swc
            * e.g. 18464_00009.xy64z16.swc
    * Folder **resample**
        * Storing .swc after affining marker points
            * e.g. 18464_00009.x128y128z32.swc
    * Folder **affine**
            * Storing original .swc
                * e.g. 18464_00009.x128y128z32_affine
    * Folder **stps**
        * Storing final version .swc
            * e.g. 

In [94]:
brain_regis

y_initial x_initial z_initial y_after x_after z_after  \
ID                                                                
17109         35000     22793     10533     547     356     659   
17302         54600     34412      9847     853     537     615   
17541         54600     39328     11073     426     307     346   
17542         54600     34748     11333     426     271     354   
17545         54600     35989     10750     853     562     671   
17781         54600     34722     11021     426     271     344   
17782         28000     45000     11786     437     703     736   
17787         24275     36400     11478     189     284     358   
17788         27188     35000     11120     212     273     347   
18047         28000     36000     11561     218     281     361   
18052         28303     35000     10923     221     273     341   
18053         25600     37000     12500     200     289     390   
18453         24400     39200     10538     190     306     329   
18454         26298     35000     11041     411     547     690   
18455         23376     40600     10892     182     317     340   
18457         24275     37800     10955     189     295     342   
18458         25376     40600     10333     198     317     322   
18459         22333     36400     11264     174     284     352   
18464         26352     35000     10431     205     273     326   
18465         26000     36000     10301     203     281     321   
18864         35001     27299     10392     547     427     650   
18869         16915     29400     10923     132     229     341   
236174        36400     23814     12061     568     372     755   
236174           NA        NA        NA      NA      NA      NA   
18455_new     23376     40600     10892     182     317     340   

                            ration     flip flip_axis x_downsample  \
ID                                                                   
17109                      xy64z16        0        NA      64.0253   
17302                      xy64z16        0        NA      64.0819   
17541                     xy128z32        0        NA      128.104   
17542              x128.221y128z32        0        NA      128.221   
17545                      xy64z16        0        NA      64.0374   
17781           x128.125y128.17z32        0        NA      128.125   
17782                      xy64z16        1         Y      64.0114   
17787            x128.17y128.44z32        0        NA      128.169   
17788           128.205y128.245z32        1         Y      128.205   
18047           x128.114y128.44z32        1         Y      128.114   
18052          x128.205y128.068z32        0        NA      128.205   
18053              x128.028y128z32        1         Y      128.028   
18453             x128.1y128.42z32        1         Y      128.105   
18454                      xy64z16        1         Y      63.9854   
18455      x128.075y128.440z32.035        1         Y      128.076   
18457          x128.136y128.439z32        1         Y      128.136   
18458          x128.076y128.162z32        0        NA      128.076   
18459          x128.169y128.351z32        0        NA      128.169   
18464          x128.205y128.546z32        0        NA      128.205   
18465          x128.114y128.079z32        0        NA      128.114   
18864                      xy64z16  unknown        NA      63.9321   
18869          x128.384y128.144z32        0        NA      128.384   
236174           x64.085y64.016z16  unknown        NA      64.0161   
236174                          NA       NA        NA           NA   
18455_new                  xy64z16        0        NA      128.076   

          y_downsample z_downsample  
ID                                   
17109          63.9854      15.9833  
17302          64.0094      16.0114  
17541          128.169      32.0029  
17542          128.169      32.0141  
17545          64.0094      16.0209  
17781          128.169      32.0378  
17782

In [103]:


brain_regis ['x_initial'] 

ID
17109        22793.0
17302        34412.0
17541        39328.0
17542        34748.0
17545        35989.0
17781        34722.0
17782        45000.0
17787        36400.0
17788        35000.0
18047        36000.0
18052        35000.0
18053        37000.0
18453        39200.0
18454        35000.0
18455        40600.0
18457        37800.0
18458        40600.0
18459        36400.0
18464        35000.0
18465        36000.0
18864        27299.0
18869        29400.0
236174       23814.0
18455_new    40600.0
Name: x_initial, dtype: float64

### Assure the corresponding four folders have been created

In [64]:
path = '/home/penglab/Desktop/Warp_swc_Step_test'
os.listdir(path )

['first_downsample',
 'average_template_25_u8_xpad.v3draw',
 'readme.docx',
 'third_warp_swc',
 'second_affine',
 '18464']

In [74]:
for brain_id in os.listdir(path ):
    print(brain_id)

first_downsample
average_template_25_u8_xpad.v3draw
readme.docx
third_warp_swc
second_affine
18464


In [87]:
 os.path.join( os.path.join(path, subpath),'affine')

'/home/penglab/Desktop/Warp_swc_Step_test/18464/affine'

In [115]:
import os
path = '/home/penglab/Desktop/Warp_swc_Step_test'
for brain_id in os.listdir(path ):
    if not (brain_id) in list(map(str, brain_regis.index.tolist())):
        print("Skip file " + str(brain_id.split('/')[-1]) )
        continue

    folder_affine =  os.path.join(os.path.join(path, subpath),'affine')
    exist_affine = os.path.exists(folder_affine)
    if not exist_affine:      
        print('Creat new folder'+ str(folder_affine))
        os.makedirs(folder_affine)  

    folder_ori =  os.path.join(os.path.join(path, subpath),'ori')
    exist_ori = os.path.exists(folder_ori)
    if not exist_ori:         
        print('Creat new folder '+ str(folder_ori))
        os.makedirs(folder_ori)  

    folder_resample =  os.path.join(os.path.join(path, subpath),'resample')
    exist_resample = os.path.exists(folder_resample )
    if not exist_resample:   
        print('Creat new folder '+ str(folder_resample))
        os.makedirs(folder_resample)  

    folder_stps =  os.path.join(os.path.join(path, subpath),'stps')
    exist_stps = os.path.exists(folder_stps)
    if not exist_stps:   
        print('Creat new folder '+ str(folder_stps))
        os.makedirs(folder_stps)  
    x_ds_size = brain_regis.loc[int(brain_id),'x_downsample']
    y_ds_size = brain_regis.loc[int(brain_id),'y_downsample']
    z_ds_size = brain_regis.loc[int(brain_id),'z_downsample']

Skip file first_downsample
Skip file average_template_25_u8_xpad.v3draw
Skip file readme.docx
Skip file third_warp_swc
Skip file second_affine


In [77]:
y_ds_size 

128.54634146341462

In [78]:
extension

'.swc'

# Step 1: Downsample

In [122]:
    for iter_swc in os.listdir(folder_ori):
        if not iter_swc.endswith(("swc", "SWC")):
            continue
        (neuron_info,extension) = os.path.splitext(iter_swc)
        temp_brain = int(neuron_info.split('_')[0])
        print('Neuron information: ' + str(neuron_info))
        #print(temp_brain)
        path_old  = os.path.join(folder_ori, iter_swc)
        path_new  = os.path.join(folder_resample, neuron_info)+ '.' + 'x' + str(round(x_ds_size,2)) + 'y' + str(round(y_ds_size,2)) + 'z'+ str(round(z_ds_size,2)) + str(extension)
        n_skip = 0
        with open(path_old, "r") as f:
            for line in f.readlines():
                line = line.strip()
                if line.startswith("#"):
                    n_skip += 1
                else:
                    break
        names = ["##n", "type", "x", "y", "z", "r", "parent"]
        swc = pd.read_csv(path_old, index_col=0, skiprows=n_skip, sep=" ",
                  usecols=[0, 1, 2, 3, 4, 5, 6],
                  names=names
                  )
        swc.loc[:,'x'] = swc.loc[:,'x']/round(x_ds_size,1)
        swc.loc[:,'y'] = swc.loc[:,'y']/round(y_ds_size,1)
        swc.loc[:,'z'] = swc.loc[:,'z']/round(z_ds_size,1)
        assert(max( swc.loc[:,'x'])) < brain_regis.loc[int(brain_id),'x_after'], " x has exceeds the boundary"
        assert(max( swc.loc[:,'y'])) < brain_regis.loc[int(brain_id),'y_after'], " y has exceeds the boundary"
        assert(max( swc.loc[:,'z'])) < brain_regis.loc[int(brain_id),'z_after'], " z has exceeds the boundary"
        swc.to_csv(path_new, sep=" ")
        print('Saving current .swc to ' + str(path_new) + '\n')

Neuron information: 18464_00009.xy64z16
Saving current .swc to /home/penglab/Desktop/Warp_swc_Step_test/18464/resample/18464_00009.xy64z16.x128.21y128.55z32.0.swc



# Step 2: Affine

In [125]:
max( swc.loc[:,'x'])

104.1875

# Step 3: Warp

## Requirement
* Assume inside the main folder, each brain has its separrate folder, which includes
    * Folder **affine**
        * Storing original .swc
            * e.g. 17781.ano.apo
    * Folder **ori**
        * Storing resampled .swc
            * e.g. 17781.s.ano.apo
    * Folder **resample**
        * Storing .swc after affining marker points
            * e.g. 17781.s.affine.ano.apo
    * Folder **stps**
        * Storing final version .swc
            * e.g. 17781.s.a.warp.ano.apo